### ISSUE TITLE
Hessian vector product for stax neural network loss

### ISSUE THREAD

**Issue by AlOa** [google/jax#1345]
Hi,
suppose I define a simple  neural network using stax 
```python
net_init, net_apply = serial(
        Dense(N1),
        Tanh,
        Dens(1)
)
in_shape = (-1, 1)
out_shape, net_params = net_init(rng, in_shape)
```
and a loss function

```python
def loss(params, net, x, y):
       f = net(params, x)
       return np.mean((f-y)**2)
```
I want a function for the hessian vector product for the loss ( hessian with respect to network paramameters).
I tried different implementations but without success.

I need it because usually for small network I use scipy L-BFGS-B, but I want to try also the newton conjugate gradient.
                    Alberto

**Comment by sschoenholz** [comment]
Here's a function that computes hessian-vector products thanks to Justin Gilmer (@jmgilmer).

```python
from jax.tree_util import tree_map, tree_multimap
def hvp(loss, params, batch, v):
  """Computes the hessian vector product Hv.

  This implementation uses forward-over-reverse mode for computing the hvp.

  Args:
    loss: function computing the loss with signature
      loss(params, batch).
    params: pytree for the parameters of the model.
    batch:  A batch of data. Any format is fine as long as it is a valid input
      to loss(params, batch).
    v: pytree of the same structure as params.

  Returns:
    hvp: array of shape [num_params] equal to Hv where H is the hessian.
  """

  loss_fn = lambda x: loss(x, batch)
  return jvp(grad(loss_fn), [params], [v])[1]


def _tree_sum(tree_left, tree_right):
  """Computes tree_left + tree_right."""
  def f(x, y):
    return x + y
  return tree_multimap(f, tree_left, tree_right)


def _tree_zeros_like(tree):
  def f(x):
    return np.zeros_like(x)
  return tree_map(f, tree)
```

**Comment by AlOa** [comment]
Thanks!

      AlOa

**Comment by MInner** [comment]
@sschoenholz could you please elaborate on why do we need two helper functions `_tree_sum` and `_tree_zeros_like` considering that (at least on the surface level) they seem to not be used in the `hpv` implementation?

**Comment by sschoenholz** [comment]
Haha, I think you're right, they aren't necessary.
